app/src/main/java/com/keylesspalace/tusky/service/SendTootService.kt

```diff
@@ -0,0 +1,245 @@
+import com.keylesspalace.tusky.util.StringUtils
+import dagger.android.AndroidInjection
+import kotlinx.android.parcel.Parcelize
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import java.util.*
+import javax.inject.Inject
+
+class SendTootService: Service(), Injectable {
+
+    @Inject
+    lateinit var mastodonApi: MastodonApi
+    @Inject
+    lateinit var accountManager: AccountManager
+
+    private lateinit var saveTootHelper: SaveTootHelper
+
+    private val tootsToSend = mutableMapOf<Int, TootToSend>()
+    private val timer = Timer()
+
+
+    override fun onCreate() {
+        AndroidInjection.inject(this)
+        saveTootHelper = SaveTootHelper(TuskyApplication.getDB().tootDao(), this)
+        super.onCreate()
+    }
+
+    override fun onBind(intent: Intent): IBinder? {
+        return null
+    }
+
+
+    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
+
+        val tootToSend = intent.getParcelableExtra<TootToSend>(KEY_TOOT)
+
+        if(tootToSend == null) {
+            throw IllegalStateException("SendTootService started without $KEY_TOOT extra")
+        }
+
+        val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+
+        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val channel = NotificationChannel(CHANNEL_ID, getString(R.string.send_toot_notification_channel_name), NotificationManager.IMPORTANCE_LOW)
+            notificationManager.createNotificationChannel(channel)
+
+        }
+
+        var notificationText = tootToSend.warningText
+        if(notificationText.isBlank()) {
+            notificationText = tootToSend.text
+        }
+
+        val builder = NotificationCompat.Builder(this, CHANNEL_ID)
+                .setSmallIcon(R.drawable.ic_notify)
+                .setContentTitle(getString(R.string.send_toot_notification_title))
+                .setContentText(notificationText)
+                .setProgress(1, 0, true)
+                .setOngoing(true)
+                .setColor(ContextCompat.getColor(this, R.color.primary))
+
+        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_DETACH)
+        startForeground(notificationId, builder.build())
+
+        tootsToSend[notificationId] = tootToSend
+        sendToot(notificationId)
+
+        notificationId--
+
+
+        return Service.START_STICKY
+
+    }
+
+
+    private fun sendToot(tootId: Int) {
+
+        // when tootToSend == null, sending has been canceled
+        val tootToSend = tootsToSend[tootId] ?: return
+
+        // when account == null, user has logged out, cancel sending
+        val account = accountManager.getAccountById(tootToSend.accountId)
+
+        if(account == null) {
+            tootsToSend.remove(tootId)
+            return
+        }
+
+        tootToSend.retries++
+
+        mastodonApi.createStatus(
+                "Bearer " + account.accessToken,
```

Comment:
Why do you pass token here directly when you could leave it to interceptor?

Reply:
because otherwise the account can be switched while the status is in the retry-loop and the status will be sent with an account different from the one the user expects

Reply:
Oh crap, it's probably my design mistake, we should have passed something like account id as the header, not the domain...
But then it wouldn't work for places like auth...
uuuuugh
I don't really like the idea of passing both everywhere but it seems it's the only flexible way?
Or we could teach interceptor about both domain and account id. The former would have higher priority and they probably won't be used in one request anyway.